.github/workflows/reusable-workflow.yml

```diff
@@ -0,0 +1,115 @@
+name: Deploy to Environment
+
+on:
+  workflow_call:
+    inputs:
+      env:
+        description: "The name of environment (eg. STAGING, PRODUCTION, SAUCE-LABS)"
+        required: true
+        type: string
+      env-url:
+        description: "The URL of the environment"
+        required: true
+        type: string
+      artifact-name:
+        description: "The name of the artifact to download"
+        required: true
+        type: string
+
+jobs:
+  test:
+    strategy:
+      matrix:
+        browser: [firefox]
+    runs-on: ubuntu-latest
+    environment:
+      name: ${{ inputs.env }}
+      url: ${{ inputs.env-url }}
+
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 1
+
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.12'
+
+      - name: Install required dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y google-chrome-stable
+          sudo apt-get install -y firefox
+
+      - name: Install uv
+        uses: astral-sh/setup-uv@v5
+
+      - name: Verify repository contents
+        run: ls -la
+
+      - name: Set up virtual environment and install dependencies
+        run: |
+          uv venv -p 3.11 
+          python -m pip install --upgrade pip 
+          source .venv/bin/activate
+          uv pip install -r requirements.txt
+          ls -la
+
+      - name: Install Firefox
+        uses: browser-actions/setup-firefox@v1
+
+      - name: Verify browser installation
+        run: |
+          firefox --version
+
+      - name: Create Logs Directory
+        run: mkdir -p latest_logs/errors
+
+      - name: Set Environment Variables for Test Execution
+        env:
+          ENV_URL: ${{ vars.ENV_URL }}
+          ENV_NAME: ${{ inputs.env }}
+        run: |
+          echo "Running in $ENV_NAME environment with URL $ENV_URL"
+
+      - name: Run Tests Based on Environment
```

Comment:
You could probably still merge these three blocks ("Run Tests Based On Environment", "Run Tests for Production" and "Run Tests for Sauce Labs") into one if you define the `STAGING_URL`, `PRODUCTION_URL`, and `SAUCE_LABS_URL` variables in the environments simply as `ENV_URL`.

If the saucelabs-specific secrets are not defined they will just be empty variables if you put them in `env` the way you do now, so you don't need to worry about setting dummy variables for the other two environments.
```
env:
          SAUCE_USERNAME: ${{ secrets.SAUCE_USERNAME }}
          SAUCE_ACCESS_KEY: ${{ secrets.SAUCE_ACCESS_KEY }}
```

I'll leave some comments on this job that apply to the other two as well, but if you merge them into one you only need to process the comments once anyway :-)

Reply:
That's the cool thing: in github environments you can have the same variable with a different value in different environments. Depending on the `environment` you set for the job you'll get the value from the different environments.

Dummy variables wouldn't be needed, you don't have to worry about the saucelabs variables not being defined for the staging and production environments, because if anything tries to use them they'll just be empty values.

As for running locally: you could have a script that exports the same variables with the correct values per environment and source that?

Feel free to ping on Teams if I misunderstood your comment :-)